File: LeetCodeTasks/31_next-permutation/solutions/java/solution.java
class Solution {
    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length <= 1) return;
        int i = nums.length - 2;
        
        // Find the first decreasing element
        while (i >= 0 && nums[i] >= nums[i + 1]) i--;
        
        if (i >= 0) {
            int j = nums.length - 1;
            // Find the element larger than the found element
            while (nums[j] <= nums[i]) j--;
            // Swap them
            swap(nums, i, j);
        }
        
        // Reverse the elements after the initially found element
        reverse(nums, i + 1, nums.length - 1);
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }
}

File: LeetCodeTasks/31_next-permutation/solutions/cpp/solution.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    void nextPermutation(std::vector<int>& nums) {
        int n = nums.size(), k, l;
        for (k = n - 2; k >= 0; k--) {
            if (nums[k] < nums[k + 1]) {
                break;
            }
        }
        if (k < 0) {
            std::reverse(nums.begin(), nums.end());
        } else {
            for (l = n - 1; l > k; l--) {
                if (nums[l] > nums[k]) {
                    break;
                }
            }
            std::swap(nums[k], nums[l]);
            std::reverse(nums.begin() + k + 1, nums.end());
        }
    }
};
